### ISSUE TITLE
Admin reset password link broken

### ISSUE THREAD

**Issue by kassner** [magento/magento2#35667]
### Preconditions and environment

- Magento 2.4.4
- Multiple websites and stores

### Steps to reproduce

1. Enable settings Add Store Code to Urls
2. Create 2 new website with each 1 store view
3. Create new admin user
5. Click `Forgot your password?` link in the admin login page;
6. Input your e-mail and submit;
7. Click the link in the email.

### Expected result

Form to reset the password

### Actual result

404 not found

### Additional information

Link is generated as `http://magento.local/admin/admin_123456/admin/auth/resetpassword/key/xxxxxxxxxx/?id=1&token=yyyyyyyy`, but it should be `http://magento.local/admin_123456/admin/auth/resetpassword/key/xxxxxxxxxx/?id=1&token=yyyyyyyy` (notice the extra `admin/` right after the domain name).

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by kassner** [comment]
@magento give me 2.4-develop instance

**Comment by kassner** [comment]
Reverting the changes from https://github.com/magento/magento2/commit/f19611c1bc8b7a18a03ef8098adc36e8d62a7b1f makes it work again.

**Comment by kassner** [comment]
Maybe because my installation has a custom admin path? We observe that in my case there is an “admin” before the custom path, when it shouldn’t.

**Comment by engcom-Lima** [comment]
Hi @kassner 

I have changed the custom path of the admin. Followed all your steps but I am able to reset password successfully. Link which I have received is below:
 **http://local.magento2.com/shubham/admin/auth/resetpassword/key/*******?id=4&token=*********
Please find below screeshot for your reference:
![image](https://user-images.githubusercontent.com/60198022/176125411-5eaa9c63-321a-4766-9289-29f6d8fe49ac.png)

**Comment by ledian-hymetllari** [comment]
HI @kassner have you checked under admin -> stores -> configurations -> ADVANCED -> admin -> admin base url, to check if you have added any custom URL from admin?

**Comment by engcom-Lima** [comment]
Hi @kassner 
We are closing this issue as there has been no latest update on the same. Kindly reopen / create new issue if you are still facing any issues.
Thank you

**Comment by anvanza** [comment]
https://github.com/magento give me 2.4-develop instance

**Comment by anvanza** [comment]
@magento give me 2.4-develop instance

**Comment by anvanza** [comment]
Steps to reproduce:

1. Enable settings Add Store Code to Urls
2. Create 2 new website with each 1 store view
3. Create new admin user
4. Reset password request for that new admin user

The email is now incorrect.

@kassner Could you verify is you have this "Add Store Code to Urls" enabled?

**Comment by anvanza** [comment]
@engcom-Lima , Could you check this please?

**Comment by Kamina92** [comment]
we solved the same problem creating a plugin for Magento\User\Model\Notificator

with this workaround 

`/**
     * Workaround to disable store code in reset password link 
     * 
     * @param Notificator $subject
     * @param callable $proceed
     * @param UserInterface $user
     * @return bool
     */
    public function aroundSendForgotPassword(
        Notificator $subject, 
        callable $proceed, 
        UserInterface $user
    ) {

        $store = $this->storeManager->getStore(Store::DEFAULT_STORE_ID);
        $store->setDisableStoreInUrl(true);

        try {
            $this->sendNotification(
                'admin/emails/forgot_email_template',
                [
                    'username' => $user->getFirstName().' '.$user->getLastName(),
                    'user' => $user,
                    'store' => $store
                ],
                $user->getEmail(),
                $user->getFirstName().' '.$user->getLastName()
            );
        } catch (LocalizedException $exception) {
            throw new NotificatorException(
                __($exception->getMessage()),
                $exception
            );
        }

        return true;
    }`

**Comment by engcom-Lima** [comment]
Hi @kassner @

**Comment by engcom-Lima** [comment]
:heavy_check_mark: Issue confirmed

Issue got reproduced in **2.4-develop** branch.

**Description:** Admin password reset link is not working.

**Pre-requisite:**

- Fresh magento `2.4-develop` should be installed.
- Multiple Website should be created

**Steps to reproduce:**

1. Enable settings Add Store Code to Urls
2. Create 2 new website with each 1 store view
3. Create new admin user
4. Reset password request for that new admin user

**Expected result:** Admin should be able to reset the password.
**Actual result:**  In my case,Admin is not able to reset password as link directing to login page.
**Observation:** Extra admin is added in the reset link.
`http://local.magento2.com/admin/admin/admin/auth/resetpassword/key/5aae6561fe4f1bf1d15ecaefe3646bc2f87afb6cbcff4f39ba1eadc96dc0b09b/?id=5&token=a88NTVfKBSqUM5yLcvLUMAiLeRDRRdjb`

**Screenshots:**
![image](https://user-images.githubusercontent.com/60198022/191255951-720f2f34-558c-4391-9f90-5b3b72b5d967.png)

**Logs:**
[logs.txt](https://github.com/magento/magento2/files/9607617/logs.txt)

Hence,confirming the issue.

Thanks

**Comment by bruno-blackbird** [comment]
We applied another workaround for this. It will solve all the cases where a store is passed directly to the url resolver.

```
diff --git a/Model/Url.php b/Model/Url.php
index 8948961..c1bf6f0 100644
--- a/Model/Url.php
+++ b/Model/Url.php
@@ -447,6 +447,13 @@ class Url extends \Magento\Framework\Url implements \Magento\Backend\Model\UrlIn
                     'data' => ['code' => 'admin', 'force_disable_rewrites' => false, 'disable_store_in_url' => true],
                 ]
             );
+        } else {
+            /**
+             * Patch admin password url incorrect with setting "web/url/use_store" = 1
+             * Force option to disable store in url from url.
+             * Issue : https://github.com/magento/magento2/issues/35667
+             */
+            $this->_scope->setData("disable_store_in_url", true);
         }
         return $this->_scope;
     }
```

**Comment by tarekfellah** [comment]
Is there a fix for this issue?

**Comment by RakeshJesadiya** [comment]
You can fix this issue without any patches also.
Just modify the one line from the vendor/magento/module-user/view/adminhtml/email/password_reset_confirmation.html email template,

From the Store URL, Remove **_admin_** from the **store url="admin/auth/resetpassword**

`{{store url="admin/auth/resetpassword/" _query_id=$user.user_id _query_token=$user.rp_token _nosid=1}}`

Replace line no 23 with the given line,

`{{store url="auth/resetpassword/" _query_id=$user.user_id _query_token=$user.rp_token _nosid=1}}`

**_### This is the way to create a patch of email templates._**

If you don't require to create a patch,
Just create a new Forgot Admin Password template from the Backend,  **Marketing -> Communication -> Email Template**
Create a new template by clicking on the **Add New Template** button,

Now Load Default Template from the **Magento_User -> Forgot Admin Password**

After Loading, just modify the above line from the email template and save the new template.

Just Assigned this email template from the **Stores -> Configuration -> Advanced -> Admin -> Admin User Emails,**

From the Dropdown, Forgot Password Email Template you need to choose the template and Save Config.

**Comment by rudy1976s** [comment]
I followed RakeshJesadiya solution, but even if I cleared the cache and whatever possible, is still sending the wrong path leading me to a 404

**Comment by dwolfuk40** [comment]
> bruno

Where do i find this file to edit??

**Comment by bruno-blackbird** [comment]
@dwolfuk40 The file is present in module "magento/module-backend". You should find it in "vendor/magento/module-backend/Model/Url.php" line 447.
You can apply the patch by following this tutorial : https://devdocs.magento.com/guides/v2.3/comp-mgr/patching.html#custom-patches

**Comment by 0franco** [comment]
Based on @bruno-blackbird workaround, I've created another patch to avoid touch any other area. Already tested on v2.4.6

`diff --git a/vendor/magento/module-user/Model/Notificator.php b/vendor/magento/module-user/Model/Notificator.php
index 3e36cd13..e452de3d 100644
--- a/vendor/magento/module-user/Model/Notificator.php
+++ b/vendor/magento/module-user/Model/Notificator.php
@@ -111,7 +111,7 @@ class Notificator implements NotificatorInterface
                     'user' => $user,
                     'store' => $this->storeManager->getStore(
                         Store::DEFAULT_STORE_ID
-                    )
+                    )->setData("disable_store_in_url", true)
                 ],
                 $user->getEmail(),
                 $user->getFirstName().' '.$user->getLastName()
`

Above patch only add the flag to the specific function that trigger the email.

**Comment by tschirmer** [comment]
Getting this on 2.4.5-p2 as well. Any feedback on getting any of these patches in the core?

**Comment by PTuijtelaars-ISAAC** [comment]
Also experiencing this problem on 2.4.5-p2

**Comment by GamerHegi64** [comment]
I have the same issue on 2.4.6-p1

**Comment by andrew-miguel** [comment]
I have the same issue on 2.4.5-p1.  Is this a bug? Can somebody confirm?

**Comment by rajnikantdas84** [comment]
@RakeshJesadiya ,

Sorry to say but your solution is not working.

**Comment by rajnikantdas84** [comment]
Hi @kassner and Team,

I am also facing the similar issue and reset password link for admin is not working if we have custom admin name.
Magento version is 2.4.6-1

so if we keep the admin frontname "admin" which is default one then reset password link generated as below and working.
https://my.local/admin/admin/admin/auth/resetpassword/key/7714b67d3b933a2c8472bc6eed93b81e98d12cb474bf45fafb12619c72829a95/?id=13&token=1oW9NQMiRJFiFsT8SUytqNtuybnPzWcg

But when we modify the frontname as "myadmin" then we get following reset password link which is not working:
https://my.local/admin/myadmin/admin/auth/resetpassword/key/7714b67d3b933a2c8472bc6eed93b81e98d12cb474bf45fafb12619c72829a95/?id=13&token=1oW9NQMiRJFiFsT8SUytqNtuybnPzWcg

If we remove the additional "admin" before our custom frontname "myadmin" then url is working.

We tried the fix mentioned by the @bruno-blackbird and that is working fine, but we wanted to know if there are any official Magento patches available for this or not or if there are any other way to fix this.

Thanks

**Comment by rjslegall** [comment]
Hey guys!

We are facing the same issue with Magento version 2.4.5-p1.
Some claim that the path is not working.
Does anyone have another solution?

When will Adobe comment on the serious problem?

**Comment by elioermini** [comment]
We are facing this issue as well. Our setup is:

- Magento 2.4.6-p3
- add store code for urls enabled
- custom backend frontName enabled in env.php

**Comment by elioermini** [comment]
We applied the PR https://github.com/magento/magento2/pull/36280 creating a patch and the email link works fine, as well as admin area, however I notice now some admin path are now _customAdminUrl/**admin**/rest_of_path_ and others just _customAdminUrl/rest_of_path_ :

```
Index: vendor/magento/module-backend/Model/Url.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/vendor/magento/module-backend/Model/Url.php b/vendor/magento/module-backend/Model/Url.php
--- a/vendor/magento/module-backend/Model/Url.php	
+++ b/vendor/magento/module-backend/Model/Url.php	(date 1699871741208)
@@ -440,13 +440,17 @@
      */
     protected function _getScope()
     {
+        $scopeData = ['code' => 'admin', 'force_disable_rewrites' => false, 'disable_store_in_url' => true];
+
         if (!$this->_scope) {
             $this->_scope = $this->_storeFactory->create(
                 [
                     'url' => $this,
-                    'data' => ['code' => 'admin', 'force_disable_rewrites' => false, 'disable_store_in_url' => true],
+                    'data' => $scopeData,
                 ]
             );
+        } else {
+            $this->_scope->addData($scopeData);
         }
         return $this->_scope;
     }

```

Running the patch with:

` patch -p1 -N < path/of/patch.patch`

but cannot guarantee that this can cause side effects. We use this setup as posted above:

> * Magento 2.4.6-p3
> * add store code for urls enabled
> * custom backend frontName enabled in env.php

**Comment by thomas-kl1** [comment]
> Based on @bruno-blackbird workaround, I've created another patch to avoid touch any other area. Already tested on v2.4.6
> 
> ```
> diff --git a/vendor/magento/module-user/Model/Notificator.php b/vendor/magento/module-user/Model/Notificator.php
> index 3e36cd13..e452de3d 100644
> --- a/vendor/magento/module-user/Model/Notificator.php
> +++ b/vendor/magento/module-user/Model/Notificator.php
> @@ -111,7 +111,7 @@ class Notificator implements NotificatorInterface
>                      'user' => $user,
>                      'store' => $this->storeManager->getStore(
>                          Store::DEFAULT_STORE_ID
> -                    )
> +                    )->setData("disable_store_in_url", true)
>                  ],
>                  $user->getEmail(),
>                  $user->getFirstName().' '.$user->getLastName()
> ```
> 
> Above patch only add the flag to the specific function that trigger the email.

Why there is no PR with this change yet?

**Comment by thomas-kl1** [comment]
Hi all, please find the official patch provided by support:

[ACSD-48570_2.4.5.patch](https://github.com/magento/magento2/files/14278424/ACSD-48570_2.4.5.patch)

**Comment by hostep** [comment]
Looks like that patch mentioned above got included in Magento 2.4.7: https://github.com/magento/magento2/commit/face8387fe48034d981d07efd45612ab2950f460

So I guess we can close this issue now?

**Comment by engcom-Delta** [comment]
Hi @kassner ,

Thank you for reporting and collaboration.
Verified the issue on Magento 2.4-develop and the issue is not reproducible. Kindly refer the screenshots

**Steps to reproduce**
1. Enable settings Add Store Code to Urls
![image](https://github.com/user-attachments/assets/afb18a48-d45b-4a9c-8806-e96302fc674f)
2. Create 2 new website with each 1 store view
3. Create new admin user
4. Click Forgot your password? link in the admin login page;
5. Input your e-mail and submit;
6. Click the link in the email.
![image](https://github.com/user-attachments/assets/4fd85b8a-d864-477f-8f7a-b8e16d9bd617)
Observe , user is able to reset password
Kindly recheck the issue on latest Magento 2.4-develop instance and Please let us know if you are still facing any issue. 
Thanks.
